cmd/net/src/lib.rs

```diff
@@ -0,0 +1,444 @@
+    // - Read the number of entries in the MAC table
+    // - Loop over the table that many times, reading entries
+    let value = humility_cmd_hiffy::hiffy_call(
+        hubris,
+        core,
+        context,
+        &op_mac_count,
+        &[],
+    )?;
+    let mac_count = match value {
+        Ok(v) => {
+            if let Value::Base(Base::U32(v)) = v {
+                v
+            } else {
+                bail!("Got bad reflected value: expected U32, got {:?}", v);
+            }
+        }
+        Err(e) => {
+            bail!("Got error: {}", e);
+        }
+    };
+
+    // Due to HIF limitations (lack of Expand16 / Collect16), we're going to
+    // limit ourselves to 255 MAC table entries.
+    let (mac_count, clipped) = if mac_count > u8::MAX as u32 {
+        (u8::MAX, true)
+    } else {
+        (mac_count as u8, false)
+    };
+
+    println!("Reading {} MAC addresses...", mac_count);
+
+    let op = IdolOperation::new(hubris, "Net", "read_ksz8463_mac", None)?;
+    let funcs = context.functions()?;
+    let send = funcs.get("Send", 4)?;
+
+    use hif::*;
+    let label = Target(0);
+
+    let ops = vec![
+        Op::Push32(op.task.task()),
+        Op::Push16(op.code),
+        Op::Push(0), // LSB of u16
+        Op::Push(0), // Dumy counter byte
+        Op::Label(label),
+        // BEGIN LOOP
+        Op::Drop,
+        Op::Push(0), // MSB of u16
+        Op::Push(2), // Payload length
+        Op::Push(8), // Return size
+        Op::Call(send.id),
+        Op::DropN(3), // Drop everything until the LSB
+        // Loop iteration is now at the top of the stack
+        Op::Push(1), // Prepare to increment
+        Op::Add,     // i = i + 1
+        Op::Push(mac_count),
+        Op::BranchGreaterThan(label),
+        // END LOOP
+        Op::DropN(3),
+        Op::Done,
+    ];
+
+    let results = context.run(core, ops.as_slice(), None)?;
+    let results = results
+        .into_iter()
+        .map(move |r| humility_cmd_hiffy::hiffy_decode(hubris, &op, r))
+        .collect::<Result<Vec<Result<_, _>>>>()?;
+
+    let mut mac_table: BTreeMap<u16, Vec<[u8; 6]>> = BTreeMap::new();
+    for r in results {
+        if let Ok(r) = r {
+            let s = r.as_struct()?;
+            assert_eq!(s.name(), "KszMacTableEntry");
+            let port = s["port"].as_base().unwrap().as_u16().unwrap();
+            let mut mac = [0; 6];
+            for (i, m) in s["mac"].as_array().unwrap().iter().enumerate() {
+                mac[i] = m.as_base().unwrap().as_u8().unwrap()
+            }
+            if mac == [0; 6] && port == 0xFFFF {
+                println!("Skipping empty MAC address");
+            } else {
+                mac_table.entry(port).or_default().push(mac);
+            }
+        } else {
+            // Log the error but keep going for other entries in the table
+            println!("Got error result: {:?}", r);
+        }
+    }
+    println!(" {} |        {}", "PORT".bold(), "MAC".bold());
```

Comment:
I wonder if it would be worth it to put our table printing into a library? (Not needed, more thinking long term)

Reply:
Yeah, or maybe switching to something like [`tabled`](https://crates.io/crates/tabled)...